sql/core/src/main/scala/org/apache/spark/sql/execution/r/MapPartitionsRWrapper.scala

```diff
@@ -25,6 +25,21 @@ import org.apache.spark.sql.Row
 import org.apache.spark.sql.types.{BinaryType, StructField, StructType}
 
 /**
+ * A function wrapper that applies the given R function to each partition of each group.
+ */
+private[sql] case class MapGroupsRWrapper(
```

Comment:
yes, key should be a list of grouping columns. Key encoder is of RowEncoder. Update RRunner.compute() to allowing passing a Row which contains the key grouping columns. In gapply mode, RRunner passes the key columns to R worker by something like this:
```
dataOut.writeInt(key.length)
(0 until key.length). map { SerDe.writeObject(key(_).asInstanceOf[Object]) }
```